### PR TITLE
melt with custom variable columns using variable_table attribute

### LINKED ISSUE #2551: Possibility to provide group labels in melt `variable` column when melting a list of columns simultaneously
Right now, there is a loss of information when one melts a list of columns simultaneously since the `variable` column only gives the number of the group (as a factor by default), not a name.

It would be great if the functionality of melting a list of columns simultaneously would be extended such that it is possible to provide labels for the factor in the `variable`    ###column. I would suggest an additional argument, e.g. `variable.labels`  which makes it possible to provide custom labels.

`#` `Minimal reproducible example`

    library(data.table)
    
    DT <- data.table(val_loss = 1:2,
                     val_acc = 3:4,
                     loss = 5:6,
                     acc = 7:8,
                     id = c("A", "B"))
    
    DT_melted <- 
      melt(DT,
           measure.vars = list(c("loss", "acc"), c("val_loss", "val_acc")),
    # the following line shows what I think would be ideal
    #      variable.labels = c("group 1", "group 2"),
           value = c("training", "validation"))
    
    # the following step should become obsolete by providing the labels during the melt call above
    DT_melted[, variable := factor(variable, levels = 1:2, labels = c("group 1", "group 2"))]

`#` `Output of sessionInfo()`

    R version 3.3.3 (2017-03-06)
    Platform: x86_64-w64-mingw32/x64 (64-bit)
    Running under: Windows 7 x64 (build 7601) Service Pack 1
    
    locale:
    [1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
    [4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    
    
    attached base packages:
    [1] stats     graphics  grDevices utils     datasets  methods   base     
    
    other attached packages:
    [1] data.table_1.10.4-3
    
    loaded via a namespace (and not attached):
    [1] tools_3.3.3

I know that the R Version is not up to date but at my Company it takes a while until we get the most recent version.
 
Sorry, for not labeling the issue. Somehow I could not figure out, how to set them.

### LINKED ISSUE #2575: Properly handling the "variable" column after melting unbalanced wide data
In reference to [my comments](https://github.com/Rdatatable/data.table/issues/2551#issuecomment-358670658) at #2551, using `patterns` with `melt` on unbalanced wide or panel data results in an incorrect molten dataset.

Here's a minimal example, where "a" values are missing "beta" times, and "b" values are missing "alpha" times:

```
library(data.table)
set.seed(2334)
DT <- data.table(
  a_alpha = rnorm(3), a_gamma = rnorm(3),
  b_beta = rnorm(3), b_gamma = rnorm(3),
  id = c(1:3)
)
```
`melt` just sees that there are two sets of variables:

```
melt(DT, measure.vars = patterns("a_", "b_"))
#    id variable     value1       value2
# 1:  1        1 -0.1183107  1.682018318
# 2:  2        1  1.2370906 -0.573611132
# 3:  3        1  0.8088209 -0.057320032
# 4:  1        2 -0.7656264 -0.706428227
# 5:  2        2 -0.5919939  0.001899857
# 6:  3        2  0.5279071  1.063851211
```

---------------------

Expected behaviour:

```
melt(DT, "id")[, c("stub", "var") := tstrsplit(variable, "_")][, dcast(.SD, id + var ~ stub)]
#    id   var          a            b
# 1:  1 alpha -0.1183107           NA
# 2:  1  beta         NA  1.682018318
# 3:  1 gamma -0.7656264 -0.706428227
# 4:  2 alpha  1.2370906           NA
# 5:  2  beta         NA -0.573611132
# 6:  2 gamma -0.5919939  0.001899857
# 7:  3 alpha  0.8088209           NA
# 8:  3  beta         NA -0.057320032
# 9:  3 gamma  0.5279071  1.063851211
```

### LINKED ISSUE #3396: melt.data.table should offer `variable` to match on the name, rather than the number
Currently, using `patterns` makes the variable column an integer, which is both a bit awkward ...

``` r
library(data.table)
DT <- 
  data.table(x = 1:5, 
             y_No = 101:105,
             `y_$`= 1011:1015,
             z_No = 201:205,
             `z_$` = 301:305)
DT_melt <- 
  melt.data.table(DT, 
                  id.vars = "x",
                  measure.vars = patterns("Dollar" = "\\$$", 
                                          "Number" = "No$"))
DT_melt
#>     x variable Dollar Number
#>  1: 1        1   1011    101
#>  2: 2        1   1012    102
#>  3: 3        1   1013    103
#>  4: 4        1   1014    104
#>  5: 5        1   1015    105
#>  6: 1        2    301    201
#>  7: 2        2    302    202
#>  8: 3        2    303    203
#>  9: 4        2    304    204
#> 10: 5        2    305    205
# Intended
DT_melt[, variable_new := c("y", "z")[variable]][]
#>     x variable Dollar Number variable_new
#>  1: 1        1   1011    101            y
#>  2: 2        1   1012    102            y
#>  3: 3        1   1013    103            y
#>  4: 4        1   1014    104            y
#>  5: 5        1   1015    105            y
#>  6: 1        2    301    201            z
#>  7: 2        2    302    202            z
#>  8: 3        2    303    203            z
#>  9: 4        2    304    204            z
#> 10: 5        2    305    205            z
```

... and potentially error-prone, e.g.

``` r
library(data.table)
DT <- 
  data.table(x = 1:5, 
             y_No = 101:105,
             `y_$`= 1011:1015,
             z_No = 201:205,
             `z_$` = 301:305,
             `w1_$` = 401:405,
             w2_No = 501:505)
melt.data.table(DT, 
                id.vars = "x",
                measure.vars = patterns("Dollar" = "\\$$", 
                                        "Number" = "No$"))
#>     x variable Dollar Number
#>  1: 1        1   1011    101
#>  2: 2        1   1012    102
#>  3: 3        1   1013    103
#>  4: 4        1   1014    104
#>  5: 5        1   1015    105
#>  6: 1        2    301    201
#>  7: 2        2    302    202
#>  8: 3        2    303    203
#>  9: 4        2    304    204
#> 10: 5        2    305    205
#> 11: 1        3    401    501    # incorrect
#> 12: 2        3    402    502    # incorrect
#> 13: 3        3    403    503    # incorrect
#> 14: 4        3    404    504    # incorrect
#> 15: 5        3    405    505    # incorrect
```

<sup>Created on 2019-02-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Can offer a PR if this is a worthwhile feature. Otherwise can just attempt the bugfix.

### PR DESCRIPTION
**Pull request by tdhock**
Closes #2551
Closes #2575
Closes #3396 
Background: there are a bunch of issues  #3396 #2575 #2551  involving melt on a data table with column names that each have more than one piece of information e.g. 

1. Petal.Length, Sepal.Length, Petal.Width, ... (part, dimension)
2. a_1, b_1, a_2, b_2, ... (letter, number)
3. sex_child1, age_child1, sex_child2, age_child2, ... (feature, number) 

In these situations we would like output columns that have names shown in parentheses above, but with current melt the variable column is either the full column name (if there is one output value column),

```r
remotes::install_github("Rdatatable/data.table@melt-custom-variable")
#> Skipping install of 'data.table' from a github remote, the SHA1 (c02fa9e8) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(data.table)
options(datatable.print.class=TRUE)
DT <- data.table(id=1, a_1=10, b_2=21, a_2=20)
melt(DT, measure.vars=patterns("_"))
#>       id variable value
#>    <num>   <fctr> <num>
#> 1:     1      a_1    10
#> 2:     1      b_2    21
#> 3:     1      a_2    20
```

 or an integer (if there are multiple output value columns),

```r
melt(DT, measure.vars=patterns(a="a", b="b"))
#>       id variable     a     b
#>    <num>   <fctr> <num> <num>
#> 1:     1        1    10    21
#> 2:     1        2    20    NA
```

Comparison: tidyr::pivot_longer can be used with names_sep/names_pattern/names_transform arguments to get a more useful molten/tall output table in these cases.

Proposal: The goal of this PR is to fix these issues, and provide melt feature-parity with tidyr::pivot_longer.
The solution involves a new function `measure` which lets us do this if we want a single output value column,
```r
melt(DT, measure.vars=measure(letter, number=as.integer, sep="_"))
#>       id letter number value
#>    <num> <char>  <int> <num>
#> 1:     1      a      1    10
#> 2:     1      b      2    21
#> 3:     1      a      2    20
```
and we can use the special `value.name` keyword if we want multiple output value columns,
```r
melt(DT, measure.vars=measure(value.name, number=as.integer, sep="_"))
#>       id number     a     b
#>    <num>  <int> <num> <num>
#> 1:     1      1    10    NA
#> 2:     1      2    20    21
```
Note in the code above that the output does not include the `variable` column at all. Instead the more relevant letter/number columns are output.

Some points to discuss:

- do we keep the name `measure` for this new function or do we want to call it something else? 
- `measure` function allows specification of type conversion, e.g., `as.integer` above. (similar to names_transform argument of tidyr::pivot_longer).
- do we keep the `value.name` keyword (for consistency with `value.name` argument)  or do we change it to something else to avoid confusion?
- internally this works because the `measure()` function returns a vector/list with a special `variable_table` attribute which is a data table with columns letter/number. There is new fmelt C code which recognizes this attribute and uses it to create the desired output. Is `variable_table` a good name for this attribute or shall we call it something else?
- I did a bunch of timings to compare the new code with existing data.table and tidyr solutions (see comment with figures below), and it seems like the new code is really fast.
- I had to re-write `do_patterns`, and I renamed it to `eval_with_cols` in order for it to support adding `cols` argument to `measure` as well as `patterns`. A nice new feature is that it will work with ANY user-provided function that has an argument named `cols`. e.g., 

```r
remotes::install_github("tdhock/nc@new-measure")
melt(DT, measure.vars=nc::measure(letter="[ab]", "_", number="[12]", as.integer))
#>       id letter number value
#>    <num> <char>  <int> <num>
#> 1:     1      a      1    10
#> 2:     1      b      2    21
#> 3:     1      a      2    20
melt(DT, measure.vars=nc::measure(column="[ab]", "_", number="[12]", as.integer))
#>       id number     a     b
#>    <num>  <int> <num> <num>
#> 1:     1      1    10    NA
#> 2:     1      2    20    21
```

I know this is a really big PR --- please tell me if there is anything I can do to help make it easier to review.
Thanks.

### CODE REVIEW

**Comment by tdhock** [comment]
this PR includes commits from other PRS #4720  and #4723 (these other ones are simpler and should be reviewed/merged first)

**Comment by tdhock** [comment]
In interactive testing the tests I added in tests.Rraw work fine, but the last three cause errors when I run them using test.data.table(). I'm doing melt(iris) and there seem to be extra/unexpected columns named `grp` and `Species1` in the output -- has anybody seen something like that / know of a fix? Also I try test.data.table("another_file.Rraw") with only the new tests that I added and that works fine too... is some other test affecting my new tests somehow?

**Comment by tdhock** [comment]
here is a demo of the new functionality,
```r
remotes::install_github("Rdatatable/data.table@melt-custom-variable")
#> Skipping install of 'data.table' from a github remote, the SHA1 (190e0194) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(data.table)
iris.dt = data.table(iris)
melt(iris.dt, measure.vars=data.table:::sep_list(part, column, sep=".", cols=names(iris)))
#>        Species  part Length Width
#>   1:    setosa Sepal    5.1   3.5
#>   2:    setosa Sepal    4.9   3.0
#>   3:    setosa Sepal    4.7   3.2
#>   4:    setosa Sepal    4.6   3.1
#>   5:    setosa Sepal    5.0   3.6
#>  ---                             
#> 296: virginica Petal    5.2   2.3
#> 297: virginica Petal    5.0   1.9
#> 298: virginica Petal    5.2   2.0
#> 299: virginica Petal    5.4   2.3
#> 300: virginica Petal    5.1   1.8
melt(iris.dt, measure.vars=data.table:::sep_list(column, dim, sep=".", cols=names(iris)))
#>        Species    dim Sepal Petal
#>   1:    setosa Length   5.1   1.4
#>   2:    setosa Length   4.9   1.4
#>   3:    setosa Length   4.7   1.3
#>   4:    setosa Length   4.6   1.5
#>   5:    setosa Length   5.0   1.4
#>  ---                             
#> 296: virginica  Width   3.0   2.3
#> 297: virginica  Width   2.5   1.9
#> 298: virginica  Width   3.0   2.0
#> 299: virginica  Width   3.4   2.3
#> 300: virginica  Width   3.0   1.8
melt(iris.dt, measure.vars=data.table:::sep_vec(part, dim, sep=".", cols=names(iris)))
#>        Species  part    dim value
#>   1:    setosa Sepal Length   5.1
#>   2:    setosa Sepal Length   4.9
#>   3:    setosa Sepal Length   4.7
#>   4:    setosa Sepal Length   4.6
#>   5:    setosa Sepal Length   5.0
#>  ---                             
#> 596: virginica Petal  Width   2.3
#> 597: virginica Petal  Width   1.9
#> 598: virginica Petal  Width   2.0
#> 599: virginica Petal  Width   2.3
#> 600: virginica Petal  Width   1.8
```
TODO: make it so we don't have to manually specify cols argument to new sep_* and pattern_* functions (a la patterns).

**Comment by tdhock** [comment]
TODO timings to see if this results in significant speedups.

**Comment by tdhock** [comment]
about the `grp` and `Species1` columns, they are added to `iris` in an earlier test, so the fix was to use `datasets::iris`.

**Comment by tdhock** [comment]
I changed `do_patterns` to `eval_with_cols` which now works for any function (not just patterns), so we can do
```r
remotes::install_github("Rdatatable/data.table@melt-custom-variable")
#> Skipping install of 'data.table' from a github remote, the SHA1 (ab9cbbb4) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(data.table)
iris.dt = data.table(iris)[1:2]
melt(iris.dt, measure.vars=sep_list(part, column, sep="."))
#>    Species  part Length Width
#> 1:  setosa Sepal    5.1   3.5
#> 2:  setosa Sepal    4.9   3.0
#> 3:  setosa Petal    1.4   0.2
#> 4:  setosa Petal    1.4   0.2
melt(iris.dt, measure.vars=sep_list(column, dim, sep="."))
#>    Species    dim Sepal Petal
#> 1:  setosa Length   5.1   1.4
#> 2:  setosa Length   4.9   1.4
#> 3:  setosa  Width   3.5   0.2
#> 4:  setosa  Width   3.0   0.2
melt(iris.dt, measure.vars=sep_vec(part, dim, sep="."))
#>    Species  part    dim value
#> 1:  setosa Sepal Length   5.1
#> 2:  setosa Sepal Length   4.9
#> 3:  setosa Sepal  Width   3.5
#> 4:  setosa Sepal  Width   3.0
#> 5:  setosa Petal Length   1.4
#> 6:  setosa Petal Length   1.4
#> 7:  setosa Petal  Width   0.2
#> 8:  setosa Petal  Width   0.2
```

**Comment by TysonStanley** [comment]
I don't have time right now to go through this in any depth, but the user functionality looks fantastic. I think it is intuitive and allows users to more simply access the speed/efficiency without, what I assume, any breaking changes.

**Comment by tdhock** [comment]
There are lots of conflicts in tests.Rraw due to the test numbers. At the time when this PR was initiated, the test numbers I used were new, but now they conflict with some test numbers which were added in other PRs which have been merged into master since then. Going forward what is the recommendation for choosing new test numbers in such a way as to avoid conflicts with other PRs? I checked the Contributing wiki page but I did not see any mention of this issue.

**Comment by MichaelChirico** [comment]
That's a pain point with our test numbering system, we don't really have a workaround.

But nothing needed on your end -- if that's the only conflict, it will be fixed by Matt during merge.
